File: typescript/src/integrations/openai.ts
/**
 * OpenAI integration — pre-built agent and LLM judge for OpenAI models.
 *
 * Usage:
 *   import { OpenAIAgent, openaiJudgeFn } from "evalforge/integrations/openai";
 *   import { EvalHarness, TestCase } from "evalforge";
 *   import { llmJudge } from "evalforge/scorer";
 *   import { Scorer } from "evalforge/scorer";
 *
 *   const agent = new OpenAIAgent({ model: "gpt-4o" });
 *   const scorer = new Scorer(openaiJudgeFn({ model: "gpt-4o-mini" }));
 *
 *   const harness = new EvalHarness({ agent, scorer, suiteName: "my-suite" });
 *   harness.add(new TestCase({ ... }));
 *   await harness.run();
 */

import type { LLMJudgeFn } from "../scorer.js";

export interface OpenAIAgentOptions {
  model?: string;
  apiKey?: string;
  systemPrompt?: string;
  temperature?: number;
  maxTokens?: number;
  [key: string]: unknown;
}

type MessageRole = "user" | "assistant" | "system";
interface ChatMessage { role: MessageRole; content: string; }

/**
 * A callable agent backed by an OpenAI chat completion model.
 * Requires the `openai` npm package: `npm install openai`
 */
export class OpenAIAgent {
  private opts: Required<Omit<OpenAIAgentOptions, "systemPrompt">> & { systemPrompt?: string };

  constructor(opts: OpenAIAgentOptions = {}) {
    this.opts = {
      model: opts.model ?? "gpt-4o-mini",
      apiKey: opts.apiKey ?? process.env.OPENAI_API_KEY ?? "",
      systemPrompt: opts.systemPrompt,
      temperature: (opts.temperature as number) ?? 0,
      maxTokens: (opts.maxTokens as number) ?? 1024,
    };
  }

  async __call__(input: unknown): Promise<string> {
    return this.call(input);
  }

  async call(input: unknown): Promise<string> {
    // Dynamic import so the package is optional
    const { default: OpenAI } = await import("openai").catch(() => {
      throw new Error('openai package required. Install with: npm install openai');
    });

    const client = new OpenAI({ apiKey: this.opts.apiKey });
    const messages = this.buildMessages(input);

    const response = await client.chat.completions.create({
      model: this.opts.model,
      messages,
      temperature: this.opts.temperature,
      max_tokens: this.opts.maxTokens,
    });

    return response.choices[0]?.message?.content ?? "";
  }

  private buildMessages(input: unknown): ChatMessage[] {
    const messages: ChatMessage[] = [];

    if (this.opts.systemPrompt) {
      messages.push({ role: "system", content: this.opts.systemPrompt });
    }

    if (typeof input === "string") {
      messages.push({ role: "user", content: input });
    } else if (
      input !== null &&
      typeof input === "object" &&
      "messages" in input &&
      Array.isArray((input as { messages: unknown[] }).messages)
    ) {
      messages.push(...(input as { messages: ChatMessage[] }).messages);
    } else if (Array.isArray(input)) {
      messages.push(...(input as ChatMessage[]));
    } else {
      messages.push({ role: "user", content: String(input) });
    }

    return messages;
  }
}

/**
 * Returns an async LLM judge function backed by an OpenAI model.
 *
 * Usage:
 *   import { Scorer } from "evalforge/scorer";
 *   const scorer = new Scorer(openaiJudgeFn({ model: "gpt-4o-mini" }));
 */
export function openaiJudgeFn(opts: OpenAIAgentOptions = {}): LLMJudgeFn {
  const model = opts.model ?? "gpt-4o-mini";
  const apiKey = opts.apiKey ?? process.env.OPENAI_API_KEY ?? "";

  return async (prompt: string): Promise<number> => {
    const { default: OpenAI } = await import("openai").catch(() => {
      throw new Error('openai package required. Install with: npm install openai');
    });

    const client = new OpenAI({ apiKey });
    const response = await client.chat.completions.create({
      model,
      messages: [
        {
          role: "system",
          content:
            "You are an impartial evaluator. Respond with ONLY a decimal number between 0.0 and 1.0.",
        },
        { role: "user", content: prompt },
      ],
      temperature: 0,
      max_tokens: 10,
    });

    const raw = response.choices[0]?.message?.content?.trim() ?? "0";
    const score = parseFloat(raw);
    return isNaN(score) ? 0 : Math.max(0, Math.min(1, score));
  };
}

File: python/evalforge/integrations/openai.py
"""
OpenAI integration - pre-built runner and LLM judge for OpenAI models.

Usage::

    from evalforge.integrations.openai import OpenAIAgent, openai_judge_fn
    from evalforge import EvalHarness, TestCase
    from evalforge.scorer import llm_judge, Scorer

    agent = OpenAIAgent(model="gpt-4o")
    scorer = Scorer(llm_judge_fn=openai_judge_fn(model="gpt-4o-mini"))

    harness = EvalHarness(agent=agent, suite_name="my-suite", scorer=scorer)
    harness.add(TestCase(
        input={"messages": [{"role": "user", "content": "What is 2+2?"}]},
        expected_output="4",
        scoring=llm_judge(threshold=0.8),
    ))
    result = harness.run()
"""

from __future__ import annotations

import logging
import os
from typing import Any, Callable, Optional

logger = logging.getLogger(__name__)


class OpenAIAgent:
    """
    A simple agent that calls an OpenAI chat completion model.

    The agent accepts input in one of two forms:
    - A string (converted to a single user message)
    - A dict with a ``messages`` key (passed directly to the API)

    Returns the content of the first assistant message.
    """

    def __init__(
        self,
        model: str = "gpt-4o-mini",
        api_key: Optional[str] = None,
        system_prompt: Optional[str] = None,
        temperature: float = 0.0,
        max_tokens: int = 1024,
        **extra_kwargs: Any,
    ):
        self.model = model
        self.api_key = api_key or os.environ.get("OPENAI_API_KEY")
        self.system_prompt = system_prompt
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.extra_kwargs = extra_kwargs

    async def __call__(self, input_data: Any) -> str:
        try:
            from openai import AsyncOpenAI
        except ImportError:
            raise ImportError(
                "openai package required for OpenAIAgent. "
                "Install with: pip install openai"
            )

        client = AsyncOpenAI(api_key=self.api_key)

        messages = self._build_messages(input_data)

        response = await client.chat.completions.create(
            model=self.model,
            messages=messages,
            temperature=self.temperature,
            max_tokens=self.max_tokens,
            **self.extra_kwargs,
        )

        return response.choices[0].message.content or ""

    def _build_messages(self, input_data: Any) -> list[dict]:
        messages = []

        if self.system_prompt:
            messages.append({"role": "system", "content": self.system_prompt})

        if isinstance(input_data, str):
            messages.append({"role": "user", "content": input_data})
        elif isinstance(input_data, dict) and "messages" in input_data:
            messages.extend(input_data["messages"])
        elif isinstance(input_data, list):
            messages.extend(input_data)
        else:
            messages.append({"role": "user", "content": str(input_data)})

        return messages


def openai_judge_fn(
    model: str = "gpt-4o-mini",
    api_key: Optional[str] = None,
) -> Callable:
    """
    Returns an async LLM judge function backed by an OpenAI model.

    The judge receives the prompt and returns a float in [0, 1].

    Usage::

        from evalforge.scorer import Scorer
        scorer = Scorer(llm_judge_fn=openai_judge_fn("gpt-4o-mini"))
    """

    async def judge(prompt: str, expected: Any, actual: Any) -> float:
        try:
            from openai import AsyncOpenAI
        except ImportError:
            raise ImportError("pip install openai")

        resolved_key = api_key or os.environ.get("OPENAI_API_KEY")
        client = AsyncOpenAI(api_key=resolved_key)

        response = await client.chat.completions.create(
            model=model,
            messages=[
                {
                    "role": "system",
                    "content": (
                        "You are an impartial evaluator. "
                        "Respond with ONLY a decimal number between 0.0 and 1.0."
                    ),
                },
                {"role": "user", "content": prompt},
            ],
            temperature=0.0,
            max_tokens=10,
        )

        raw = (response.choices[0].message.content or "0").strip()
        try:
            score = float(raw)
            return max(0.0, min(1.0, score))
        except ValueError:
            logger.warning("LLM judge returned non-numeric value: %r", raw)
            return 0.0

    return judge
